{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d94f462c",
   "metadata": {},
   "source": [
    "# Easy Manipulation of Genetic Variant Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f17ba94c",
   "metadata": {},
   "source": [
    "For analysis of genetic variants, single nucleotide polymorphism (SNP) information is widely used. A SNP corresponds to a nucleotide position on the genome where some degree of variation has been observed in a population, with each individual have one of two possible alleles at that position on each of a pair of chromosomes. The two alleles are often distinguished as a reference allele (REF, allele on the [reference genome](https://en.wikipedia.org/wiki/Reference_genome)) and an alternate allele (ALT).  The most widely used repository for SNP information is [dbSNP](https://www.ncbi.nlm.nih.gov/snp/), where each SNP is indexed by an identifier beginning with `rs` (\"rsID\"). \n",
    "\n",
    "The _genotypes_ of each sample for each variant is commonly coded in:\n",
    "\n",
    "| value | genotype |\n",
    "|:---:|:---:|\n",
    "| 0 | homozygous REF |\n",
    "| 1 | heterozyguous REF/ALT |\n",
    "| 2 | homozygous ALT |\n",
    "\n",
    "Sometimes, when the genotype has uncertainties, it is represented in _dosage_ after imputation. Given the posterior probabilities of each genotype, dosage is computed as \n",
    "$$\\mathrm{dosage = 0 \\cdot Prob(REF/REF) + 1 \\cdot Prob(REF/ALT) + 2 \\cdot Prob(ALT/ALT)}$$\n",
    "and have values in $[0, 2]$. \n",
    "\n",
    "\n",
    "\n",
    "Often, the data formats for genetic variants include\n",
    "- Information of $m$ samples (identifier, sex, phenotypes, etc.)\n",
    "- Information $n$ SNPs (identifier, chromosome, position, REF/ALT alleles, etc.)\n",
    "- A $m \\times n$ table or matrix containing the observed allelic type at $n$  positions for $m$ individuals.\n",
    "\n",
    "A common type of analysis for this data is [genome-wide association studies (GWAS)](https://en.wikipedia.org/wiki/Genome-wide_association_study), often testing a statistical hypothesis variant by variant. Significance of each SNP is assessed by some type of regression:\n",
    "$$\n",
    "\\mathrm{trait âˆ¼ SNP + age + sex +  principal\\;components + other\\;covariates }\n",
    "$$\n",
    "\n",
    "In this workshop, we learn four widely-used file types for genetic variants and how to manipulate them in Julia.  \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02bc249",
   "metadata": {},
   "source": [
    "We focus on accessing genotype information variant by varint, as it is a common workflow for a GWAS-based application. We learn how do this and try to compute some simple properties of SNPs, such as minor allele frequencies (MAF)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eaa3c89",
   "metadata": {},
   "source": [
    "## Genetic Variant File Formats\n",
    "\n",
    "\n",
    "### Variant Call Format (`.vcf`)\n",
    "Text-based format is the most intuitive to represent genetic variants. It is the most flexible format to include diverse information on the samples and variants.\n",
    "\n",
    "\n",
    "- __Pros__: Highly flexible, easy to parse and interpret\n",
    "- __Cons__: Large file size\n",
    "\n",
    "With the scale of recent genetic data, with data set like UK Biobank having near a million subjects and millions of variants, the storage needed for storing a raw VCF file is prohibitively huge. A common approach to remedy this is by storing the data in a compressed form (such as `.gz` file) and decompress it at analysis time as a stream. However, there is a trade-off between stored file size and time needed for decompression.\n",
    "\n",
    "- __Pros__: Reasonable file size with flexible representation\n",
    "- __Cons__: Takes too long to decompress\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "761f5d82",
   "metadata": {},
   "source": [
    "#### Basic usage: `VCFTools.jl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e74e60ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.1\n",
      "##INFO=<ID=LDAF,Number=1,Type=Float,Description=\"MLE Allele Frequency Accounting for LD\">\n",
      "##INFO=<ID=AVGPOST,Number=1,Type=Float,Description=\"Average posterior probability from MaCH/Thunder\">\n",
      "##INFO=<ID=RSQ,Number=1,Type=Float,Description=\"Genotype imputation quality from MaCH/Thunder\">\n",
      "##INFO=<ID=ERATE,Number=1,Type=Float,Description=\"Per-marker Mutation rate from MaCH/Thunder\">\n",
      "##INFO=<ID=THETA,Number=1,Type=Float,Description=\"Per-marker Transition rate from MaCH/Thunder\">\n",
      "##INFO=<ID=CIEND,Number=2,Type=Integer,Description=\"Confidence interval around END for imprecise variants\">\n",
      "##INFO=<ID=CIPOS,Number=2,Type=Integer,Description=\"Confidence interval around POS for imprecise variants\">\n",
      "##INFO=<ID=END,Number=1,Type=Integer,Description=\"End position of the variant described in this record\">\n",
      "##INFO=<ID=HOMLEN,Number=.,Type=Integer,Description=\"Length of base pair identical micro-homology at event breakpoints\">\n",
      "##INFO=<ID=HOMSEQ,Number=.,Type=String,Description=\"Sequence of base pair identical micro-homology at event breakpoints\">\n",
      "##INFO=<ID=SVLEN,Number=1,Type=Integer,Description=\"Difference in length between REF and ALT alleles\">\n",
      "##INFO=<ID=SVTYPE,Number=1,Type=String,Description=\"Type of structural variant\">\n",
      "##INFO=<ID=AC,Number=.,Type=Integer,Description=\"Alternate Allele Count\">\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total Allele Count\">\n",
      "##ALT=<ID=DEL,Description=\"Deletion\">\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "##FORMAT=<ID=DS,Number=1,Type=Float,Description=\"Genotype dosage from MaCH/Thunder\">\n",
      "##FORMAT=<ID=GL,Number=.,Type=Float,Description=\"Genotype Likelihoods\">\n",
      "##INFO=<ID=AA,Number=1,Type=String,Description=\"Ancestral Allele, ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/pilot_data/technical/reference/ancestral_alignments/README\">\n",
      "##INFO=<ID=AF,Number=1,Type=Float,Description=\"Global Allele Frequency based on AC/AN\">\n",
      "##INFO=<ID=AMR_AF,Number=1,Type=Float,Description=\"Allele Frequency for samples from AMR based on AC/AN\">\n",
      "##INFO=<ID=ASN_AF,Number=1,Type=Float,Description=\"Allele Frequency for samples from ASN based on AC/AN\">\n",
      "##INFO=<ID=AFR_AF,Number=1,Type=Float,Description=\"Allele Frequency for samples from AFR based on AC/AN\">\n",
      "##INFO=<ID=EUR_AF,Number=1,Type=Float,Description=\"Allele Frequency for samples from EUR based on AC/AN\">\n",
      "##INFO=<ID=VT,Number=1,Type=String,Description=\"indicates what type of variant the line represents\">\n",
      "##INFO=<ID=SNPSOURCE,Number=.,Type=String,Description=\"indicates if a snp was called when analysing the low coverage or exome alignment data\">\n",
      "##reference=GRCh37\n",
      "##reference=GRCh37\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tHG00096\tHG00097\tHG00099\tHG00100\tHG00101\tHG00102\tHG00103\tHG00104\tHG00106\tHG00108\tHG00109\tHG00110\tHG00111\tHG00112\tHG00113\tHG00114\tHG00116\tHG00117\tHG00118\tHG00119\tHG00120\tHG00121\tHG00122\tHG00123\tHG00124\tHG00125\tHG00126\tHG00127\tHG00128\tHG00129\tHG00130\tHG00131\tHG00133\tHG00134\tHG00135\tHG00136\tHG00137\tHG00138\tHG00139\tHG00140\tHG00141\tHG00142\tHG00143\tHG00146\tHG00148\tHG00149\tHG00150\tHG00151\tHG00152\tHG00154\tHG00155\tHG00156\tHG00158\tHG00159\tHG00160\tHG00171\tHG00173\tHG00174\tHG00176\tHG00177\tHG00178\tHG00179\tHG00180\tHG00182\tHG00183\tHG00185\tHG00186\tHG00187\tHG00188\tHG00189\tHG00190\tHG00231\tHG00232\tHG00233\tHG00234\tHG00235\tHG00236\tHG00237\tHG00238\tHG00239\tHG00240\tHG00242\tHG00243\tHG00244\tHG00245\tHG00246\tHG00247\tHG00249\tHG00250\tHG00251\tHG00252\tHG00253\tHG00254\tHG00255\tHG00256\tHG00257\tHG00258\tHG00259\tHG00260\tHG00261\tHG00262\tHG00263\tHG00264\tHG00265\tHG00266\tHG00267\tHG00268\tHG00269\tHG00270\tHG00271\tHG00272\tHG00273\tHG00274\tHG00275\tHG00276\tHG00277\tHG00278\tHG00280\tHG00281\tHG00282\tHG00284\tHG00285\tHG00306\tHG00309\tHG00310\tHG00311\tHG00312\tHG00313\tHG00315\tHG00318\tHG00319\tHG00320\tHG00321\tHG00323\tHG00324\tHG00325\tHG00326\tHG00327\tHG00328\tHG00329\tHG00330\tHG00331\tHG00332\tHG00334\tHG00335\tHG00336\tHG00337\tHG00338\tHG00339\tHG00341\tHG00342\tHG00343\tHG00344\tHG00345\tHG00346\tHG00349\tHG00350\tHG00351\tHG00353\tHG00355\tHG00356\tHG00357\tHG00358\tHG00359\tHG00360\tHG00361\tHG00362\tHG00364\tHG00366\tHG00367\tHG00369\tHG00372\tHG00373\tHG00375\tHG00376\tHG00377\tHG00378\tHG00381\tHG00382\tHG00383\tHG00384\tHG00403\tHG00404\tHG00406\tHG00407\tHG00418\tHG00419\tHG00421\tHG00422\tHG00427\tHG00428\n",
      "22\t20000086\trs138720731\tT\tC\t100\tPASS\tAC=7;RSQ=0.8454;AVGPOST=0.9983;AA=T;AN=2184;LDAF=0.0040;THETA=0.0001;VT=SNP;SNPSOURCE=LOWCOV;ERATE=0.0003;AF=0.0032;AFR_AF=0.01\tGT:DS:GL\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.04,-1.05,-5.00\t0/0:0.000:-0.07,-0.85,-5.00\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.06,-0.87,-5.00\t0/0:0.000:-0.03,-1.14,-5.00\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.23,-0.45,-1.28\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.11,-0.65,-4.40\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.06,-0.91,-5.00\t0/0:0.000:-0.18,-0.47,-2.54\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.01,-1.74,-5.00\t0/0:0.000:-0.00,-3.66,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.00,-2.53,-5.00\t0/0:0.000:-0.09,-0.73,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.00,-3.11,-5.00\t0/0:0.000:-0.06,-0.89,-5.00\t0/0:0.000:-0.09,-0.71,-4.10\t0/0:0.000:-0.11,-0.65,-4.40\t0/0:0.000:-0.18,-0.47,-2.34\t0/0:0.000:-0.22,-0.45,-1.32\t0/0:0.000:-0.02,-1.29,-5.00\t0/0:0.000:-0.03,-1.15,-5.00\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.00,-3.34,-5.00\t0/0:0.000:-0.12,-0.61,-3.19\t0/0:0.000:-0.11,-0.67,-4.40\t0/0:0.000:-0.05,-0.99,-5.00\t0/0:0.000:-0.18,-0.48,-2.15\t0/0:0.000:-0.01,-1.47,-5.00\t0/0:0.000:-0.10,-0.67,-3.62\t0/0:0.000:-0.03,-1.14,-5.00\t0/0:0.000:-0.09,-0.73,-4.40\t0/0:0.000:-0.07,-0.84,-4.40\t0/0:0.000:-0.18,-0.48,-2.46\t0/0:0.000:-0.0292813,-1.18575,-5\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.67,-5.00\t0/0:0.000:-0.18,-0.47,-2.40\t0/0:0.000:-0.03,-1.25,-5.00\t0/0:0.000:-0.11,-0.66,-3.44\t0/0:0.000:-0.09,-0.73,-4.70\t0/0:0.000:-0.0418663,-1.03687,-4.39794\t0/0:0.000:-0.08,-0.79,-3.14\t0/0:0.000:-0.00,-2.30,-5.00\t0/0:0.000:-0.00,-2.54,-5.00\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.06,-0.86,-5.00\t0/0:0.000:-0.09,-0.71,-4.70\t0/0:0.000:-0.01,-1.49,-5.00\t0/0:0.000:-0.01,-1.88,-5.00\t0/0:0.000:-0.09,-0.71,-4.70\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.10,-0.67,-4.40\t0/0:0.000:-0.01,-1.51,-5.00\t0/0:0.000:-0.02,-1.40,-5.00\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.00,-2.02,-5.00\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.050:-0.18,-0.47,-2.73\t0/0:0.000:-0.17,-0.49,-2.97\t0/0:0.000:-0.10,-0.68,-4.40\t0/0:0.000:-0.05,-0.99,-5.00\t0/0:0.000:-0.12,-0.62,-3.38\t0/0:0.000:-0.00,-2.06,-5.00\t0/0:0.000:-0.16,-0.51,-2.66\t0/0:0.000:-0.11,-0.64,-4.22\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.01,-1.64,-5.00\t0/0:0.000:-0.00,-2.85,-5.00\t0/0:0.000:-0.02,-1.38,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.0311436,-1.15989,-5\t0/0:0.000:-0.36,-0.42,-0.73\t0/0:0.000:-0.01,-1.88,-5.00\t0/0:0.000:-0.05,-0.92,-5.00\t0/0:0.000:-0.03,-1.16,-5.00\t0/0:0.000:-0.04,-1.04,-5.00\t0/0:0.000:-0.13,-0.59,-5.00\t0/0:0.000:-0.02,-1.36,-5.00\t0/0:0.000:-0.16,-0.51,-2.36\t0/0:0.000:-0.02,-1.31,-5.00\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.00,-4.40,-5.00\t0/0:0.000:-0.03,-1.16,-5.00\t0/0:0.000:-0.09,-0.73,-3.70\t0/0:0.000:-0.19,-0.47,-1.77\t0/0:0.000:-0.00,-3.32,-5.00\t0/0:0.000:-0.17,-0.51,-2.00\t0/0:0.000:-0.00,-2.17,-5.00\t0/0:0.000:-0.00,-2.91,-5.00\t0/0:0.000:-0.10,-0.71,-4.10\t0/0:0.000:-0.03,-1.12,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.67,-5.00\t0/0:0.000:-0.00,-2.09,-5.00\t0/0:0.000:-0.04,-1.09,-5.00\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.02,-1.41,-5.00\t0/0:0.000:-0.10,-0.69,-3.80\t0/0:0.000:-0.01,-1.54,-5.00\t0/0:0.000:-0.03,-1.16,-5.00\t0/0:0.000:-0.09,-0.73,-4.70\t0/0:0.000:-0.09,-0.74,-4.70\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.08,-0.78,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.10,-0.67,-4.40\t0/0:0.000:-0.01,-1.71,-5.00\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.02,-1.26,-5.00\t0/0:0.000:-0.04,-1.10,-5.00\t0/0:0.000:-0.02,-1.27,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.47,-5.00\t0/0:0.000:-0.00,-2.00,-5.00\t0/0:0.000:-0.10,-0.67,-4.22\t0/0:0.050:-0.18,-0.47,-2.34\t0/0:0.000:-0.05,-1.00,-5.00\t0/0:0.000:-0.11,-0.65,-3.85\t0/0:0.000:-0.10,-0.68,-4.70\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.08,-0.76,-5.00\t0/0:0.000:-0.19,-0.47,-2.14\t0/0:0.000:-0.00,-1.99,-5.00\t0/0:0.000:-0.18,-0.47,-2.46\t0/0:0.000:-0.09,-0.74,-4.40\t0/0:0.450:-0.05,-0.94,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.10,-0.69,-4.70\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.18,-0.47,-2.34\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.06,-0.88,-5.00\t0/0:0.000:-0.02,-1.41,-5.00\t0/0:0.000:-0.06,-0.88,-5.00\t0/0:0.000:-0.18,-0.47,-1.95\t0/0:0.000:-0.19,-0.46,-2.17\t0/0:0.000:-0.03,-1.13,-5.00\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.18,-0.48,-2.23\t0/0:0.000:-0.23,-0.45,-1.31\t0/0:0.000:-0.11,-0.64,-3.92\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.11,-0.66,-4.22\t0/0:0.000:-0.12,-0.61,-2.38\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.40,-0.45,-0.60\t0/0:0.000:-0.00,-2.98,-5.00\t0/0:0.000:-0.13,-0.59,-2.09\t0/0:0.000:-0.02,-1.37,-5.00\t0/0:0.000:-0.477139,-0.477113,-0.477113\t0/0:0.000:-0.04,-1.10,-5.00\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.01,-1.51,-5.00\t0/0:0.000:-0.01,-1.67,-5.00\t0/0:0.000:-0.08,-0.75,-4.40\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.10,-0.69,-4.40\t0/0:0.000:-0.12,-0.63,-3.92\t0/0:0.000:-0.01,-1.74,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.19,-0.46,-2.60\t0/0:0.000:-0.19,-0.46,-2.62\t0/0:0.000:-0.11,-0.65,-4.70\t0/0:0.000:-0.11,-0.66,-4.70\t0/0:0.050:-0.18,-0.49,-2.04\t0/0:0.050:-0.10,-0.67,-4.40\t0/0:0.000:-0.01,-1.62,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.23,-0.41,-1.51\t0/0:0.000:-0.18,-0.48,-2.18\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.10,-0.68,-4.10\t0/0:0.000:-0.03,-1.24,-5.00\t0/0:0.000:-0.18,-0.48,-2.14\n",
      "22\t20000146\trs73387790\tG\tA\t100\tPASS\tLDAF=0.0169;RSQ=0.9482;THETA=0.0004;AA=G;AN=2184;AVGPOST=0.9972;VT=SNP;SNPSOURCE=LOWCOV;AC=36;ERATE=0.0003;AF=0.02;AFR_AF=0.07;EUR_AF=0.0013\tGT:DS:GL\t0/0:0.000:-0.00,-2.68,-5.00\t0/0:0.000:-0.07,-0.82,-5.00\t0/0:0.000:-0.13,-0.60,-3.05\t0/0:0.000:-0.03,-1.24,-5.00\t0/0:0.000:-0.18,-0.47,-3.08\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.16,-0.51,-2.63\t0/0:0.000:-0.01,-1.76,-5.00\t0/0:0.000:-0.10,-0.67,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.11,-0.66,-4.40\t0/0:0.000:-0.10,-0.69,-4.70\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.18,-0.47,-2.30\t0/0:0.000:-0.00,-2.80,-5.00\t0/0:0.000:-0.00,-2.02,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.49,-5.00\t0/0:0.000:-0.01,-1.76,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.02,-1.40,-5.00\t0/0:0.000:-0.00,-2.09,-5.00\t0/0:0.000:-0.10,-0.70,-4.10\t0/0:0.000:-0.22,-0.46,-1.27\t0/0:0.000:-0.18,-0.48,-2.39\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.19,-0.47,-2.27\t0/0:0.000:-0.07,-0.85,-5.00\t0/0:0.000:-0.00,-2.53,-5.00\t0/0:0.000:-0.00,-2.83,-5.00\t0/0:0.000:-0.22,-0.46,-1.24\t0/0:0.000:-0.19,-0.46,-2.27\t0/0:0.000:-0.10,-0.68,-4.40\t0/0:0.000:-0.09,-0.73,-4.22\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.15,-0.55,-2.64\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.08,-0.76,-4.70\t0/0:0.000:-0.01,-1.49,-5.00\t0/0:0.000:-0.06,-0.86,-5.00\t0/0:0.000:-0.029681,-1.18006,-5\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.18,-0.48,-2.33\t0/0:0.000:-0.10,-0.70,-4.70\t0/0:0.000:-0.00,-2.04,-5.00\t0/0:0.000:-0.03,-1.24,-5.00\t0/0:0.000:-0.10,-0.69,-4.40\t0/0:0.000:-0.0843997,-0.755377,-3.00877\t0/0:0.000:-0.21,-0.47,-1.33\t0/0:0.000:-0.00,-3.22,-5.00\t0/0:0.000:-0.00,-3.70,-5.00\t0/0:0.000:-0.05,-0.96,-5.00\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.01,-1.47,-5.00\t0/0:0.000:-0.01,-1.86,-5.00\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.01,-1.51,-5.00\t0/0:0.000:-0.03,-1.25,-5.00\t0/0:0.000:-0.01,-1.92,-5.00\t0/0:0.000:-0.00,-2.58,-5.00\t0/0:0.000:-0.06,-0.89,-5.00\t0/0:0.000:-0.05,-1.00,-5.00\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.01,-1.72,-5.00\t0/0:0.000:-0.00,-2.25,-5.00\t0/0:0.000:-0.02,-1.43,-5.00\t0/0:0.000:-0.18,-0.48,-2.03\t0/0:0.000:-0.18,-0.47,-2.72\t0/0:0.000:-0.09,-0.72,-4.70\t0/0:0.000:-0.18,-0.47,-2.42\t0/0:0.000:-0.19,-0.46,-2.20\t0/0:0.000:-0.24,-0.44,-1.19\t0/0:0.000:-0.01,-1.75,-5.00\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.19,-0.46,-2.27\t0/0:0.000:-0.05,-0.99,-5.00\t0/0:0.000:-0.06,-0.87,-5.00\t0/0:0.000:-0.00,-3.40,-5.00\t0/0:0.000:-0.10,-0.68,-4.70\t0/0:0.000:-0.01,-1.72,-5.00\t0/0:0.000:-0.0148341,-1.47392,-5\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.03,-1.16,-5.00\t0/0:0.000:-0.04,-1.03,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.09,-0.73,-4.40\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.02,-1.26,-5.00\t0/0:0.000:-0.03,-1.25,-5.00\t0/0:0.000:-0.19,-0.45,-2.12\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.05,-0.96,-5.00\t0/0:0.000:-0.00,-4.70,-5.00\t0/0:0.000:-0.13,-0.58,-3.06\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.17,-0.51,-2.09\t0/0:0.000:-0.00,-2.29,-5.00\t0/0:0.000:-0.38,-0.43,-0.67\t0/0:0.000:-0.00,-2.81,-5.00\t0/0:0.000:-0.00,-3.25,-5.00\t0/0:0.000:-0.22,-0.46,-1.26\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.01,-1.54,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.00,-2.04,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.10,-0.68,-3.70\t0/0:0.000:-0.09,-0.72,-5.00\t0/0:0.000:-0.01,-1.77,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.01,-1.51,-5.00\t0/0:0.000:-0.16,-0.51,-2.74\t0/0:0.000:-0.10,-0.69,-4.40\t0/0:0.000:-0.18,-0.48,-2.26\t0/0:0.000:-0.18,-0.48,-2.37\t0/0:0.000:-0.18,-0.48,-2.27\t0/0:0.000:-0.00,-2.58,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.00,-2.34,-5.00\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.10,-0.69,-4.40\t0/0:0.000:-0.08,-0.78,-4.70\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.18,-0.48,-2.36\t0/0:0.000:-0.01,-1.53,-5.00\t0/0:0.000:-0.18,-0.48,-2.25\t0/0:0.000:-0.10,-0.68,-4.70\t0/0:0.000:-0.09,-0.73,-5.00\t0/0:0.000:-0.02,-1.41,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.18,-0.47,-2.41\t0/0:0.000:-0.09,-0.73,-4.40\t0/0:0.000:-0.00,-2.00,-5.00\t0/0:0.000:-0.19,-0.46,-2.19\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.05,-0.98,-5.00\t0/0:0.000:-0.18,-0.47,-2.31\t0/0:0.000:-0.09,-0.73,-4.10\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.10,-0.67,-3.66\t0/0:0.000:-0.12,-0.63,-4.70\t0/0:0.000:-0.16,-0.51,-2.28\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.01,-1.75,-5.00\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.10,-0.68,-4.22\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.12,-0.62,-3.08\t0/0:0.000:-0.19,-0.45,-2.25\t0/0:0.000:-0.01,-1.77,-5.00\t0/0:0.000:-0.13,-0.60,-2.15\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.04,-1.05,-5.00\t0/0:0.000:-0.19,-0.46,-2.30\t0/0:0.000:-0.19,-0.46,-2.28\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.190265,-0.457324,-2.2321\t0/0:0.000:-0.23,-0.46,-1.21\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.00,-2.44,-5.00\t0/0:0.000:-0.01,-1.73,-5.00\t0/0:0.000:-0.01,-1.53,-5.00\t0/0:0.000:-0.05,-0.96,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.19,-0.46,-2.62\t0/0:0.000:-0.00,-2.25,-5.00\t0/0:0.000:-0.19,-0.46,-2.23\t0/0:0.000:-0.11,-0.65,-4.70\t0/0:0.000:-0.18,-0.46,-2.68\t0/0:0.000:-0.11,-0.66,-4.40\t0/0:0.000:-0.19,-0.46,-2.43\t0/0:0.000:-0.18,-0.49,-2.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.20,-0.45,-2.05\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.10,-0.68,-4.00\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.10,-0.69,-4.22\t0/0:0.000:-0.11,-0.65,-4.40\t0/0:0.000:-0.11,-0.67,-3.85\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\t20000199\trs183293480\tA\tC\t100\tPASS\tLDAF=0.0009;THETA=0.0004;AN=2184;AVGPOST=0.9990;VT=SNP;AA=A;RSQ=0.6274;SNPSOURCE=LOWCOV;AC=1;ERATE=0.0003;AF=0.0005;EUR_AF=0.0013\tGT:DS:GL\t0/0:0.000:-0.00,-2.04,-5.00\t0/0:0.000:-0.07,-0.82,-3.47\t0/0:0.000:-0.07,-0.83,-5.00\t0/0:0.000:-0.03,-1.12,-5.00\t0/0:0.000:-0.11,-0.64,-4.10\t0/0:0.000:-0.12,-0.62,-3.85\t0/0:0.000:-0.01,-1.47,-5.00\t0/0:0.000:-0.01,-1.54,-5.00\t0/0:0.000:-0.10,-0.70,-4.70\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.16,-0.50,-3.30\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.10,-0.70,-5.00\t0/0:0.000:-0.19,-0.46,-2.46\t0/0:0.000:-0.16,-0.51,-2.67\t0/0:0.000:-0.00,-2.57,-5.00\t0/0:0.000:-0.00,-2.85,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.00,-2.55,-5.00\t0/0:0.000:-0.00,-2.02,-5.00\t0/0:0.050:-0.48,-0.48,-0.48\t0/0:0.000:-0.23,-0.46,-1.24\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.10,-0.68,-4.40\t0/0:0.000:-0.06,-0.88,-5.00\t0/0:0.000:-0.13,-0.61,-2.23\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.26,-0.43,-1.11\t0/0:0.000:-0.04,-1.01,-5.00\t0/0:0.000:-0.12,-0.62,-3.28\t0/0:0.000:-0.00,-2.27,-5.00\t0/0:0.000:-0.00,-2.42,-5.00\t0/0:0.000:-0.04,-1.08,-5.00\t0/0:0.000:-0.06,-0.88,-5.00\t0/0:0.000:-0.04,-1.03,-5.00\t0/0:0.000:-0.10,-0.70,-4.10\t0/0:0.000:-0.18,-0.48,-2.22\t0/0:0.000:-0.02,-1.41,-5.00\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.14,-0.57,-2.12\t0/0:0.000:-0.10,-0.69,-4.70\t0/0:0.000:-0.12,-0.62,-2.28\t0/0:0.000:-0.0149779,-1.4698,-5\t0/0:0.000:-0.10,-0.70,-4.70\t0/0:0.000:-0.06,-0.89,-5.00\t0/0:0.000:-0.39,-0.23,-2.25\t0/0:0.000:-0.00,-3.40,-5.00\t0/0:0.000:-0.07,-0.86,-4.40\t0/0:0.000:-0.18,-0.48,-2.35\t0/0:0.000:-0.00967891,-1.65679,-5\t0/0:0.000:-0.22,-0.46,-1.24\t0/0:0.000:-0.00,-2.27,-5.00\t0/0:0.000:-0.00,-3.12,-5.00\t0/0:0.000:-0.01,-1.73,-5.00\t0/0:0.000:-0.02,-1.39,-5.00\t0/0:0.000:-0.10,-0.70,-4.70\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.09,-0.72,-4.10\t0/0:0.000:-0.10,-0.71,-4.70\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.04,-1.11,-5.00\t0/0:0.000:-0.03,-1.14,-5.00\t0/0:0.000:-0.00,-2.42,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.11,-0.65,-3.85\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.01,-1.72,-5.00\t0/0:0.000:-0.00,-2.08,-5.00\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.16,-0.52,-2.11\t0/0:0.000:-0.10,-0.69,-4.70\t0/0:0.000:-0.44,-0.46,-0.54\t0/0:0.000:-0.07,-0.83,-5.00\t0/0:0.000:-0.16,-0.51,-2.42\t0/0:0.000:-0.18,-0.48,-2.30\t0/0:0.000:-0.19,-0.46,-2.27\t0/0:0.000:-0.06,-0.89,-5.00\t0/0:0.000:-0.06,-0.88,-5.00\t0/0:0.000:-0.00,-3.15,-5.00\t0/0:0.000:-0.12,-0.61,-4.10\t0/0:0.000:-0.06,-0.91,-5.00\t0/0:0.000:-0.00935928,-1.67121,-5\t0/0:0.000:-0.18,-0.48,-2.13\t0/0:0.000:-0.07,-0.85,-5.00\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.03,-1.15,-5.00\t0/0:0.000:-0.09,-0.71,-4.70\t0/0:0.000:-0.19,-0.46,-2.52\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.09,-0.73,-4.10\t0/0:0.000:-0.11,-0.64,-4.40\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.19,-0.47,-1.83\t0/0:0.000:-0.00,-2.76,-5.00\t0/0:0.000:-0.09,-0.73,-3.92\t0/0:0.000:-0.15,-0.55,-2.43\t0/0:0.000:-0.18,-0.48,-1.89\t0/0:0.000:-0.00,-3.18,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.02,-1.43,-5.00\t0/0:0.000:-0.22,-0.40,-5.00\t0/0:0.000:-0.07,-0.84,-4.70\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.10,-0.69,-3.85\t0/0:0.000:-0.01,-1.84,-5.00\t0/0:0.000:-0.00,-3.07,-5.00\t0/0:0.000:-0.10,-0.69,-3.85\t0/0:0.000:-0.06,-0.91,-5.00\t0/0:0.000:-0.10,-0.68,-4.70\t0/0:0.000:-0.10,-0.69,-3.80\t0/0:0.000:-0.48,-0.18,-4.10\t0/0:0.000:-0.02,-1.47,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.18,-0.48,-2.34\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.18,-0.48,-2.37\t0/0:0.000:-0.10,-0.70,-4.40\t0/0:0.000:-0.00,-2.44,-5.00\t0/0:0.000:-0.01,-1.78,-5.00\t0/0:0.000:-0.00,-2.06,-5.00\t0/0:0.000:-0.00,-2.28,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.03,-1.24,-5.00\t0/0:0.000:-0.01,-1.51,-5.00\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.11,-0.64,-4.70\t0/0:0.000:-0.19,-0.46,-2.54\t0/0:0.000:-0.02,-1.38,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.02,-1.26,-5.00\t0/0:0.000:-0.18,-0.47,-2.47\t0/0:0.000:-0.16,-0.51,-2.43\t0/0:0.000:-0.01,-1.73,-5.00\t0/0:0.000:-0.19,-0.46,-2.79\t0/0:0.000:-0.06,-0.91,-5.00\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.18,-0.48,-2.35\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.19,-0.46,-2.72\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.16,-0.51,-2.71\t0/0:0.000:-0.01,-1.80,-5.00\t0/0:0.000:-0.10,-0.69,-3.66\t0/0:0.000:-0.10,-0.69,-4.22\t0/0:0.000:-0.23,-0.44,-1.26\t0/0:0.000:-0.19,-0.46,-2.12\t0/0:0.000:-0.04,-1.07,-5.00\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.19,-0.46,-2.20\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.11,-0.65,-3.44\t0/0:0.000:-0.05,-1.00,-5.00\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.01,-1.77,-5.00\t0/0:0.000:-0.189687,-0.458221,-2.2426\t0/0:0.000:-0.01,-1.79,-5.00\t0/0:0.050:-0.24,-0.42,-1.39\t0/0:0.000:-0.19,-0.45,-4.70\t0/0:0.000:-0.00,-2.15,-5.00\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.18,-0.48,-2.23\t0/0:0.000:-0.00,-2.11,-5.00\t0/0:0.000:-0.11,-0.66,-4.70\t0/0:0.000:-0.01,-1.59,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.18,-0.46,-2.63\t0/0:0.000:-0.11,-0.66,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.19,-0.46,-2.43\t0/0:0.000:-0.19,-0.47,-1.84\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.19,-0.46,-2.21\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.82,-5.00\t0/0:0.000:-0.05,-0.99,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.00,-2.26,-5.00\t0/0:0.000:-0.10,-0.69,-4.00\n",
      "22\t20000291\trs185807825\tG\tT\t100\tPASS\tERATE=0.0005;AVGPOST=0.9983;AA=G;AN=2184;LDAF=0.0015;VT=SNP;SNPSOURCE=LOWCOV;RSQ=0.5564;AC=2;THETA=0.0003;AF=0.0009;ASN_AF=0.0035\tGT:DS:GL\t0/0:0.000:-0.00,-2.06,-5.00\t0/0:0.000:-0.07,-0.83,-5.00\t0/0:0.000:-0.02,-1.27,-5.00\t0/0:0.000:-0.01,-1.77,-5.00\t0/0:0.000:-0.19,-0.45,-2.14\t0/0:0.000:-0.11,-0.66,-5.00\t0/0:0.000:-0.03,-1.17,-5.00\t0/0:0.000:-0.02,-1.33,-5.00\t0/0:0.000:-0.18,-0.48,-2.43\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.18,-0.46,-2.76\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.03,-1.15,-5.00\t0/0:0.000:-0.07,-0.83,-4.40\t0/0:0.000:-0.02,-1.33,-5.00\t0/0:0.000:-0.10,-0.68,-4.40\t0/0:0.000:-0.00,-2.28,-5.00\t0/0:0.000:-0.01,-1.83,-5.00\t0/0:0.000:-0.13,-0.61,-2.24\t0/0:0.000:-0.00,-3.70,-5.00\t0/0:0.000:-0.00,-2.21,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.22,-0.46,-1.25\t0/0:0.000:-0.00,-2.81,-5.00\t0/0:0.000:-0.02,-1.44,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.08,-0.79,-3.21\t0/0:0.000:-0.01,-1.72,-5.00\t0/0:0.000:-0.21,-0.46,-1.42\t0/0:0.000:-0.08,-0.79,-4.40\t0/0:0.000:-0.04,-1.06,-5.00\t0/0:0.000:-0.02,-1.42,-5.00\t0/0:0.000:-0.00,-3.85,-5.00\t0/0:0.000:-0.07,-0.83,-5.00\t0/0:0.000:-0.01,-1.80,-5.00\t0/0:0.000:-0.11,-0.66,-4.40\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.01,-1.47,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.19,-0.46,-2.54\t0/0:0.000:-0.03,-1.22,-5.00\t0/0:0.000:-0.05,-1.00,-5.00\t0/0:0.000:-0.13,-0.60,-2.19\t0/0:0.000:-0.0021071,-2.31515,-5\t0/0:0.000:-0.10,-0.68,-4.40\t0/0:0.000:-0.10,-0.68,-4.22\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.00,-3.38,-5.00\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.20,-0.45,-1.99\t0/0:0.000:-0.0193877,-1.35992,-5\t0/0:0.000:-0.02,-1.28,-5.00\t0/0:0.000:-0.00,-2.46,-5.00\t0/0:0.000:-0.00,-2.75,-5.00\t0/0:0.000:-0.18,-0.47,-2.36\t0/0:0.000:-0.10,-0.68,-4.00\t0/0:0.000:-0.02,-1.34,-5.00\t0/0:0.000:-0.00,-3.27,-5.00\t0/0:0.250:-0.03,-1.19,-5.00\t0/0:0.000:-0.07,-0.84,-5.00\t0/0:0.000:-0.01,-1.52,-5.00\t0/0:0.000:-0.11,-0.65,-3.40\t0/0:0.000:-0.10,-0.70,-4.70\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.00,-1.99,-5.00\t0/0:0.000:-0.01,-1.61,-5.00\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.10,-0.69,-4.22\t0/0:0.000:-0.01,-1.49,-5.00\t0/0:0.000:-0.01,-1.59,-5.00\t0/0:0.000:-0.10,-0.69,-3.36\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.08,-0.77,-4.10\t0/0:0.000:-0.05,-1.00,-5.00\t0/0:0.000:-0.10,-0.67,-3.80\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.06,-0.86,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.11,-0.66,-4.70\t0/0:0.000:-0.00,-2.68,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.01,-1.84,-5.00\t0/0:0.000:-0.0293556,-1.18469,-5\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.19,-0.46,-2.37\t0/0:0.000:-0.10,-0.68,-4.10\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.06,-0.91,-5.00\t0/0:0.000:-0.18,-0.48,-2.06\t0/0:0.000:-0.06,-0.87,-5.00\t0/0:0.000:-0.03,-1.13,-5.00\t0/0:0.000:-0.03,-1.24,-5.00\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.00,-3.85,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.07,-0.83,-4.40\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.00,-2.82,-5.00\t0/0:0.000:-0.08,-0.76,-3.80\t0/0:0.000:-0.00,-3.47,-5.00\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.11,-0.65,-3.59\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.02,-1.43,-5.00\t0/0:0.000:-0.00,-2.73,-5.00\t0/0:0.000:-0.06,-0.86,-4.22\t0/0:0.000:-0.14,-0.57,-2.84\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.01,-1.76,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.01,-1.79,-5.00\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.01,-1.87,-5.00\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.08,-0.79,-5.00\t0/0:0.000:-0.06,-0.91,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.00,-2.89,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.20,-0.44,-2.07\t0/0:0.000:-0.00,-2.65,-5.00\t0/0:0.000:-0.02,-1.27,-5.00\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.01,-1.79,-5.00\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.10,-0.68,-4.70\t0/0:0.000:-0.35,-0.41,-0.78\t0/0:0.000:-0.09,-0.75,-3.70\t0/0:0.000:-0.05,-0.98,-5.00\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.08,-0.77,-5.00\t0/0:0.000:-0.04,-1.03,-5.00\t0/0:0.000:-0.02,-1.47,-5.00\t0/0:0.000:-0.03,-1.16,-5.00\t0/0:0.000:-0.01,-1.57,-5.00\t0/0:0.000:-0.06,-0.89,-5.00\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.00,-2.30,-5.00\t0/0:0.000:-0.05,-0.98,-5.00\t0/0:0.000:-0.10,-0.69,-4.40\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.06,-0.87,-5.00\t0/0:0.000:-0.16,-0.51,-2.37\t0/0:0.000:-0.19,-0.46,-2.23\t0/0:0.050:-0.27,-0.34,-3.22\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.08,-0.77,-4.40\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.05,-0.99,-5.00\t0/0:0.000:-0.07,-0.82,-5.00\t0/0:0.000:-0.12,-0.61,-2.91\t0/0:0.000:-0.00,-1.98,-5.00\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.07,-0.82,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.49,-5.00\t0/0:0.000:-0.12,-0.61,-3.15\t0/0:0.000:-0.00,-2.08,-5.00\t0/0:0.000:-0.02,-1.33,-5.00\t0/0:0.000:-0.01,-1.74,-5.00\t0/0:0.050:-0.11227,-0.642561,-4.22185\t0/0:0.000:-0.22,-0.46,-1.28\t0/0:0.000:-0.10,-0.70,-4.10\t0/0:0.000:-0.10,-0.67,-3.62\t0/0:0.000:-0.03,-1.18,-5.00\t0/0:0.000:-0.03,-1.24,-5.00\t0/0:0.000:-0.10,-0.70,-4.10\t0/0:0.000:-0.01,-1.80,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.02,-1.47,-5.00\t0/0:0.000:-0.06,-0.88,-5.00\t0/0:0.000:-0.03,-1.13,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.18,-0.46,-2.69\t0/0:0.000:-0.03,-1.13,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.18,-0.48,-2.41\t0/0:0.000:-0.18,-0.47,-2.85\t0/0:0.050:-0.48,-0.48,-0.48\t0/0:0.000:-0.08,-0.76,-4.70\t0/0:0.000:-0.10,-0.69,-4.00\t0/0:0.000:-0.11,-0.65,-3.62\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.70,-5.00\t0/0:0.000:-0.00,-2.57,-5.00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\t20000428\trs55902548\tG\tT\t100\tPASS\tAC=323;AVGPOST=0.9983;AA=G;AN=2184;VT=SNP;RSQ=0.9949;LDAF=0.1473;SNPSOURCE=LOWCOV;ERATE=0.0003;THETA=0.0003;AF=0.15;ASN_AF=0.0017;AMR_AF=0.15;AFR_AF=0.31;EUR_AF=0.15\tGT:DS:GL\t1/0:1.000:-5.00,0.00,-5.00\t0/0:0.000:-0.35,-0.43,-0.73\t0/1:1.000:-1.81,-0.01,-2.95\t0/0:0.000:-0.01,-1.79,-5.00\t0/0:0.000:-0.06,-0.86,-5.00\t1/0:1.000:-0.19,-0.46,-2.18\t0/0:0.000:-0.10,-0.68,-5.00\t0/1:1.000:-4.40,-0.03,-1.12\t0/1:1.000:-5.00,-0.69,-0.10\t0/0:0.000:-0.10,-0.69,-4.70\t0/0:0.000:-0.48,-0.48,-0.48\t0/1:1.000:-5.00,-0.01,-1.77\t0/0:0.000:-0.18,-0.48,-2.57\t0/0:0.000:-0.02,-1.31,-5.00\t0/0:0.000:-0.11,-0.65,-4.70\t0/0:0.000:-0.10,-0.68,-4.70\t0/0:0.000:-0.01,-1.72,-5.00\t1/0:1.000:-5.00,0.00,-5.00\t1/0:1.000:-1.38,-0.02,-2.61\t0/1:1.000:-5.00,-1.40,-0.02\t0/0:0.000:-0.00,-2.97,-5.00\t0/0:0.000:-0.19,-0.47,-2.15\t0/0:0.000:-0.44,-0.46,-0.54\t0/0:0.000:-0.00,-2.52,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.01,-1.77,-5.00\t0/1:0.750:-0.22,-0.46,-1.26\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.21,-0.46,-1.42\t0/0:0.000:-0.19,-0.45,-2.20\t0/0:0.000:-0.12,-0.63,-3.36\t0/0:0.000:-0.00,-2.54,-5.00\t0/0:0.000:-0.00,-2.88,-5.00\t1/0:1.000:-2.10,-0.00,-4.00\t1/0:1.250:-5.00,-1.62,-0.01\t0/1:1.000:-3.06,-0.47,-0.18\t0/1:1.000:-5.00,-0.87,-0.06\t1/1:2.000:-5.00,-0.84,-0.07\t0/0:0.000:-0.05,-0.95,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.14,-0.58,-2.17\t0/0:0.000:-0.01,-1.77,-5.00\t0/0:0.000:-0.06,-0.91,-5.00\t1/0:1.000:-5,0,-5\t0/0:0.000:-0.05,-0.95,-5.00\t1/0:1.000:-4.70,-0.70,-0.10\t0/0:0.000:-0.09,-0.72,-4.70\t0/0:0.000:-0.01,-1.54,-5.00\t1/1:2.000:-5.00,-0.68,-0.10\t0/0:0.000:-0.18,-0.48,-2.33\t0/0:0.000:-0.00465443,-1.97224,-5\t0/0:0.000:-0.48,-0.48,-0.48\t1/0:1.000:-5.00,-0.93,-0.05\t0/0:0.000:-0.00,-2.82,-5.00\t1/1:2.000:-5.00,-1.67,-0.01\t0/0:0.000:-0.05,-0.95,-5.00\t1/0:1.000:-5.00,-0.87,-0.06\t0/0:0.000:-0.01,-1.78,-5.00\t0/0:0.000:-0.01,-1.83,-5.00\t0/1:1.000:-5.00,-0.00,-3.70\t0/1:0.950:-0.15,-0.53,-2.39\t0/0:0.000:-0.09,-0.71,-4.70\t0/0:0.000:-0.18,-0.48,-2.56\t0/0:0.000:-0.00,-2.60,-5.00\t0/0:0.000:-0.01,-1.49,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.04,-1.09,-5.00\t0/0:0.000:-0.00,-2.40,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.01,-1.54,-5.00\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.01,-1.84,-5.00\t0/1:1.000:-2.39,-0.30,-0.31\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.11,-0.64,-3.44\t0/0:0.000:-0.02,-1.42,-5.00\t0/0:0.000:-0.01,-1.76,-5.00\t0/0:0.000:-0.03,-1.17,-5.00\t0/1:1.000:-5.00,-0.03,-1.18\t0/0:0.000:-0.03,-1.21,-5.00\t0/0:0.000:-0.10,-0.70,-5.00\t0/1:1.000:-5.00,0.00,-5.00\t0/0:0.000:-0.01,-1.58,-5.00\t0/1:1.000:-5.00,0.00,-5.00\t1/0:1.000:-5,-5.21375e-05,-3.92082\t0/0:0.000:-0.05,-0.92,-5.00\t0/0:0.000:-0.02,-1.45,-5.00\t0/1:1.000:-5.00,-0.00,-4.22\t0/0:0.000:-0.09,-0.73,-4.22\t0/0:0.000:-0.00,-2.03,-5.00\t0/0:0.000:-0.03,-1.20,-5.00\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.18,-0.48,-2.03\t0/0:0.000:-0.01,-1.59,-5.00\t0/0:0.000:-0.37,-0.42,-0.71\t0/0:0.000:-0.02,-1.45,-5.00\t0/0:0.000:-0.00,-3.36,-5.00\t1/0:1.000:-3.85,-0.05,-0.94\t1/0:1.000:-5.00,-0.84,-0.07\t1/0:1.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.00,-2.56,-5.00\t0/0:0.000:-0.19,-0.47,-1.73\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.00,-3.74,-5.00\t0/0:0.000:-0.22,-0.46,-1.26\t1/0:1.000:-5.00,-0.00,-3.92\t0/0:0.000:-0.10,-0.69,-3.85\t0/0:0.000:-0.00,-2.76,-5.00\t0/0:0.000:-0.01,-1.70,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.06,-0.92,-5.00\t0/0:0.000:-0.01,-1.73,-5.00\t0/0:0.000:-0.05,-1.00,-5.00\t1/0:1.000:-2.32,-0.01,-1.58\t0/1:0.950:-0.09,-0.73,-4.70\t0/1:1.000:-2.25,-0.02,-1.52\t0/0:0.000:-0.01,-1.50,-5.00\t0/0:0.000:-0.10,-0.70,-4.40\t0/0:0.000:-0.05,-0.99,-5.00\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.03,-1.25,-5.00\t0/0:0.000:-0.10,-0.67,-4.10\t0/1:1.000:-5.00,-0.00,-4.22\t0/0:0.000:-0.18,-0.48,-2.02\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.00,-4.70,-5.00\t0/0:0.000:-0.03,-1.23,-5.00\t0/0:0.000:-0.05,-0.93,-5.00\t0/0:0.000:-0.03,-1.20,-5.00\t1/0:1.000:-0.37,-0.24,-2.87\t0/0:0.000:-0.05,-0.94,-5.00\t0/0:0.000:-0.18,-0.48,-2.36\t0/0:0.000:-0.18,-0.48,-1.83\t0/0:0.000:-0.18,-0.48,-2.23\t0/0:0.000:-0.01,-1.78,-5.00\t0/0:0.000:-0.06,-0.90,-5.00\t0/0:0.000:-0.07,-0.82,-5.00\t0/0:0.000:-0.10,-0.70,-4.70\t0/0:0.000:-0.01,-1.48,-5.00\t0/0:0.000:-0.02,-1.46,-5.00\t0/0:0.000:-0.18,-0.48,-2.11\t0/0:0.000:-0.10,-0.69,-4.70\t0/0:0.000:-0.18,-0.47,-2.84\t0/0:0.000:-0.03,-1.21,-5.00\t1/0:1.000:-4.40,-0.00,-4.70\t0/1:0.800:-0.03,-1.20,-5.00\t0/0:0.000:-0.05,-0.98,-5.00\t0/0:0.000:-0.10,-0.68,-3.70\t0/0:0.000:-0.10,-0.67,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.10,-0.68,-3.59\t0/0:0.000:-0.02,-1.26,-5.00\t0/0:0.000:-0.48,-0.48,-0.48\t0/0:0.000:-0.01,-1.80,-5.00\t1/0:1.000:-5.00,-0.66,-0.11\t0/0:0.000:-0.12,-0.61,-2.49\t1/1:2.000:-5.00,-1.30,-0.02\t0/0:0.000:-0.04,-1.07,-5.00\t0/0:0.000:-0.18,-0.48,-2.19\t0/0:0.000:-0.07,-0.85,-5.00\t1/0:1.000:-2.07,-0.04,-1.10\t0/1:1.000:-0.09,-0.74,-4.70\t0/1:1.000:-5.00,-0.00,-2.47\t0/0:0.000:-0.02,-1.26,-5.00\t1/0:1.000:-2.25,-0.01,-1.55\t0/0:0.000:-0.00,-3.17,-5.00\t0/0:0.000:-0.203967,-0.438136,-1.99396\t0/0:0.000:-0.12,-0.62,-3.18\t0/0:0.000:-0.03,-1.19,-5.00\t0/0:0.000:-0.01,-1.66,-5.00\t1/0:1.000:-5.00,-0.00,-3.74\t1/0:1.000:-1.43,-0.02,-4.70\t0/0:0.000:-0.05,-0.96,-5.00\t0/0:0.000:-0.09,-0.74,-4.70\t0/0:0.000:-0.48,-0.48,-0.48\t1/0:1.000:-5.00,-0.18,-0.47\t0/0:0.000:-0.17,-0.50,-2.70\t0/1:1.000:-5.00,-0.00,-3.80\t0/1:1.000:-2.22,-0.01,-1.97\t0/1:1.000:-5.00,-0.67,-0.10\t0/0:0.000:-0.21,-0.43,-2.01\t0/0:0.000:-0.19,-0.47,-1.83\t0/0:0.000:-0.10,-0.69,-4.40\t0/0:0.000:-0.11,-0.64,-4.10\t0/0:0.000:-0.18,-0.47,-2.80\t0/0:0.000:-0.01,-1.47,-5.00\t0/0:0.000:-0.22,-0.43,-1.61\t0/0:0.000:-0.02,-1.47,-5.00\t0/0:0.000:-0.01,-1.70,-5.00\t0/0:0.000:-0.05,-0.97,-5.00\t0/0:0.000:-0.02,-1.28,-5.00\n"
     ]
    }
   ],
   "source": [
    "using VCFTools\n",
    "\n",
    "fh = openvcf(\"test_vcf.vcf.gz\", \"r\")\n",
    "for l in 1:35\n",
    "    println(readline(fh))\n",
    "end\n",
    "close(fh)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95145579",
   "metadata": {},
   "source": [
    "As in typical VCF files, it has a bunch of meta-information lines, one header line, and then one line for each each marker. In this VCF, genetic data has fields GT (genotype), DS (dosage), and GL (genotype likelihood)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc5fbe6e",
   "metadata": {},
   "source": [
    "To access number of records and samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e6db307c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1356"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = nrecords(\"test_vcf.vcf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03b376e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "191"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = nsamples(\"test_vcf.vcf.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dae9c416",
   "metadata": {},
   "source": [
    "Information on samples and variants can be retrieved using the `VariantCallFormat` package:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ccf0591",
   "metadata": {},
   "source": [
    "Sample names can be retrieved by: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f73c6e3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "191-element Vector{String}:\n",
       " \"HG00096\"\n",
       " \"HG00097\"\n",
       " \"HG00099\"\n",
       " \"HG00100\"\n",
       " \"HG00101\"\n",
       " \"HG00102\"\n",
       " \"HG00103\"\n",
       " \"HG00104\"\n",
       " \"HG00106\"\n",
       " \"HG00108\"\n",
       " \"HG00109\"\n",
       " \"HG00110\"\n",
       " \"HG00111\"\n",
       " â‹®\n",
       " \"HG00383\"\n",
       " \"HG00384\"\n",
       " \"HG00403\"\n",
       " \"HG00404\"\n",
       " \"HG00406\"\n",
       " \"HG00407\"\n",
       " \"HG00418\"\n",
       " \"HG00419\"\n",
       " \"HG00421\"\n",
       " \"HG00422\"\n",
       " \"HG00427\"\n",
       " \"HG00428\""
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using VariantCallFormat\n",
    "reader = VCF.Reader(openvcf(\"test_vcf.vcf.gz\"))\n",
    "h = header(reader)\n",
    "h.sampleID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed42c536",
   "metadata": {},
   "source": [
    "Information of each variant is accessible by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae950702",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrom\tposition\tids\treference\talternative\n",
      "22\t20000086\t[\"rs138720731\"]\tT\t[\"C\"]\n",
      "22\t20000146\t[\"rs73387790\"]\tG\t[\"A\"]\n",
      "22\t20000199\t[\"rs183293480\"]\tA\t[\"C\"]\n",
      "22\t20000291\t[\"rs185807825\"]\tG\t[\"T\"]\n",
      "22\t20000428\t[\"rs55902548\"]\tG\t[\"T\"]\n",
      "22\t20000683\t[\"rs142720028\"]\tA\t[\"G\"]\n",
      "22\t20000771\t[\"rs114690707\"]\tA\t[\"C\"]\n",
      "22\t20000793\t[\"rs189842693\"]\tT\t[\"C\"]\n",
      "22\t20000810\t[\"rs147349046\"]\tC\t[\"T\"]\n",
      "22\t20000814\t[\"rs183154520\"]\tT\t[\"C\"]\n",
      "22\t20000864\t[\"rs187930998\"]\tG\t[\"A\"]\n",
      "22\t20000882\t[\"rs148068532\"]\tC\t[\"G\"]\n",
      "22\t20000950\t[\"rs1978233\"]\tT\t[\"G\"]\n",
      "22\t20000975\t[\"rs141800233\"]\tG\t[\"A\"]\n",
      "22\t20001001\t[\"rs192051979\"]\tT\t[\"C\"]\n",
      "22\t20001006\t[\"rs2079702\"]\tG\t[\"A\"]\n",
      "22\t20001016\t[\"rs183256914\"]\tC\t[\"T\"]\n",
      "22\t20001157\t[\"rs150580380\"]\tG\t[\"A\"]\n",
      "22\t20001159\t[\"rs139570132\"]\tC\t[\"T\"]\n",
      "22\t20001219\t[\"rs143369598\"]\tG\t[\"C\"]\n",
      "22\t20001333\t[\"rs5993894\"]\tC\t[\"T\"]\n",
      "22\t20001434\t[\"rs146344141\"]\tC\t[\"T\"]\n",
      "22\t20001455\t[\"rs188666449\"]\tG\t[\"A\"]\n",
      "22\t20001521\t[\"rs139601437\"]\tC\t[\"A\"]\n",
      "22\t20001587\t[\"rs71788814\"]\tCAG\t[\"C\"]\n",
      "22\t20001600\t[\"rs144217522\"]\tT\t[\"A\"]\n",
      "22\t20001655\t[\"rs192606530\"]\tG\t[\"A\"]\n",
      "22\t20001822\t[\"rs111598545\"]\tA\t[\"G\"]\n",
      "22\t20002011\t[\"rs184950746\"]\tC\t[\"T\"]\n",
      "22\t20002207\t[\"rs142461772\"]\tG\t[\"A\"]\n"
     ]
    }
   ],
   "source": [
    "reader = VCF.Reader(openvcf(\"test_vcf.vcf.gz\"))\n",
    "println(\"chrom\\tposition\\tids\\treference\\talternative\")\n",
    "cnt = 0\n",
    "for record in reader\n",
    "    println(\"$(VCF.chrom(record))\\t$(VCF.pos(record))\\t$(try VCF.id(record) catch; [\".\"] end)\\t$(VCF.ref(record))\\t$(VCF.alt(record))\")\n",
    "    cnt += 1\n",
    "    if cnt == 30\n",
    "        break\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff8e4a2",
   "metadata": {},
   "source": [
    "Each row of a VCF file represents a single variant, so it is natural to parse the genotypes or dosages variant-by-variant. The function `copy_gt!()` computes genotypes of each variant, and `copy_ds!()` computes dosages of each variant, represented by values in $[0, 2]$. \n",
    "\n",
    "\n",
    "The keyword arguments: \n",
    "- `model` defines which genotype model to use. The common choice is `:additive` for 0/1/2 encoding. \n",
    "- `impute` sets whether to impute the missing values with the mean of nonmissing values.\n",
    "- `sampleID` stores the list of sample ID.\n",
    "-`record_chr`, `record_pos`, `record_ids`, `record_ref`, and `record_ref` stores chromosome, position, identifiers, reference allele, and alternative alleles, respectively.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7c7b5f33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.2931937172774869\n",
      "0.0\n",
      "0.0\n",
      "0.015706806282722512\n",
      "0.0\n",
      "0.0\n",
      "0.005235602094240838\n",
      "0.0\n",
      "1.931937172774869\n",
      "0.0\n",
      "0.0\n",
      "1.801047120418848\n",
      "0.005235602094240838\n",
      "0.0\n",
      "0.0\n",
      "0.015706806282722512\n",
      "0.0\n",
      "0.03664921465968586\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.04712041884816754\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "using VariantCallFormat\n",
    "using Statistics\n",
    "using StatsBase\n",
    "# initialize VCF reader\n",
    "people, snps = nsamples(\"test_vcf.vcf.gz\"), nrecords(\"test_vcf.vcf.gz\")\n",
    "reader = VCF.Reader(openvcf(\"test_vcf.vcf.gz\"))\n",
    "\n",
    "# pre-allocate vector for marker data\n",
    "g = zeros(Union{Missing, Float64}, people)\n",
    "sampleID = Vector{String}(undef, people)\n",
    "rec_chr = Vector{String}(undef, 1)\n",
    "rec_pos = Vector{Any}(undef, 1)\n",
    "rec_ids = Vector{Vector{String}}(undef, 1)\n",
    "rec_ref = Vector{String}(undef, 1)\n",
    "rec_alt = Vector{Vector{String}}(undef, 1)\n",
    "\n",
    "for j = 1:30\n",
    "    copy_gt!(g, reader; model = :additive, impute = true, \n",
    "        sampleID=sampleID,\n",
    "        record_chr=rec_chr,\n",
    "        record_pos=rec_pos,\n",
    "        record_ids=rec_pos,\n",
    "        record_ref=rec_ref,\n",
    "        record_alt=rec_alt)\n",
    "    println(mean(g))\n",
    "end\n",
    "close(reader)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a97e1416",
   "metadata": {},
   "source": [
    "### PLINK 1 BED format (`.bed` + `.bim` and `.fam`)\n",
    "\n",
    "In order to rapidly process the variant information while keeping the file size small, various binary file types are devised. This format is basically a compact array of two-bit representation of genotypes. \n",
    "This is the native format of the well-celebrated large-scale GWAS tool, [PLINK 1](https://www.cog-genomics.org/plink/). \n",
    "This is suitable for high-performance applications directly dealing with genotype matrices such as [iterative hard thresholding](https://github.com/OpenMendel/MendelIHT.jl) and [admixture analysis](https://github.com/OpenMendel/OpenADMIXTURE.jl).  Major weakness of this format is that it cannot contain imputed information when there is uncertainty in genotypes.\n",
    "\n",
    "| Genotype | Plink/SnpArray |\n",
    "|:---:|:---:|\n",
    "| A1,A1 | 0x00 |\n",
    "| missing | 0x01 |\n",
    "| A1,A2 | 0x02 |\n",
    "| A2,A2 | 0x03 |\n",
    "\n",
    "\n",
    "\n",
    "- __Pros__: \n",
    "    - Small size. 1/32, 1/16, 1/4 compared to double-precision/single-precision/single character representation\n",
    "    - Fixed-width format suitable for massively accelerated computation (e.g. GPU acceleration)\n",
    "- __Cons__: \n",
    "    - Only contains hard-called genotypes (0, 1, 2, or missing), and cannot contain imputed values\n",
    "    - Need external files for variant (`.bim`) and sample (`.fam`) information\n",
    "    - REF/ALT designation unsupported\n",
    "    \n",
    "#### Basic usage: `SnpArrays.jl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa7abcb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SnpData(people: 191, snps: 1356,\n",
       "snp_info: \n",
       " Row â”‚ chromosome  snpid        genetic_distance  position  allele1  allele2\n",
       "     â”‚ String      String       Float64           Int64     String   String\n",
       "â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
       "   1 â”‚ 22          rs138720731               0.0  20000086  C        T\n",
       "   2 â”‚ 22          rs73387790                0.0  20000146  A        G\n",
       "   3 â”‚ 22          rs183293480               0.0  20000199  C        A\n",
       "   4 â”‚ 22          rs185807825               0.0  20000291  T        G\n",
       "   5 â”‚ 22          rs55902548                0.0  20000428  T        G\n",
       "   6 â”‚ 22          rs142720028               0.0  20000683  G        A\n",
       "â€¦,\n",
       "person_info: \n",
       " Row â”‚ fid        iid        father     mother     sex        phenotype\n",
       "     â”‚ Abstractâ€¦  Abstractâ€¦  Abstractâ€¦  Abstractâ€¦  Abstractâ€¦  Abstractâ€¦\n",
       "â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
       "   1 â”‚ 0          HG00096    0          0          0          -9\n",
       "   2 â”‚ 0          HG00097    0          0          0          -9\n",
       "   3 â”‚ 0          HG00099    0          0          0          -9\n",
       "   4 â”‚ 0          HG00100    0          0          0          -9\n",
       "   5 â”‚ 0          HG00101    0          0          0          -9\n",
       "   6 â”‚ 0          HG00102    0          0          0          -9\n",
       "â€¦,\n",
       "srcbed: test_bed.bed\n",
       "srcbim: test_bed.bim\n",
       "srcfam: test_bed.fam\n",
       ")"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using SnpArrays\n",
    "d = SnpData(\"test_bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27fd0e6a",
   "metadata": {},
   "source": [
    "Information of i-th sample is accessible by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5ad9e533",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<p>DataFrameRow (6 columns)</p><div class=\"data-frame\"><table class=\"data-frame\"><thead><tr><th></th><th>fid</th><th>iid</th><th>father</th><th>mother</th><th>sex</th><th>phenotype</th></tr><tr><th></th><th title=\"AbstractString\">Abstractâ€¦</th><th title=\"AbstractString\">Abstractâ€¦</th><th title=\"AbstractString\">Abstractâ€¦</th><th title=\"AbstractString\">Abstractâ€¦</th><th title=\"AbstractString\">Abstractâ€¦</th><th title=\"AbstractString\">Abstractâ€¦</th></tr></thead><tbody><tr><th>20</th><td>0</td><td>HG00119</td><td>0</td><td>0</td><td>0</td><td>-9</td></tr></tbody></table></div>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccc}\n",
       "\t& fid & iid & father & mother & sex & phenotype\\\\\n",
       "\t\\hline\n",
       "\t& Abstractâ€¦ & Abstractâ€¦ & Abstractâ€¦ & Abstractâ€¦ & Abstractâ€¦ & Abstractâ€¦\\\\\n",
       "\t\\hline\n",
       "\t20 & 0 & HG00119 & 0 & 0 & 0 & -9 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1mDataFrameRow\u001b[0m\n",
       "\u001b[1m Row \u001b[0mâ”‚\u001b[1m fid       \u001b[0m\u001b[1m iid       \u001b[0m\u001b[1m father    \u001b[0m\u001b[1m mother    \u001b[0m\u001b[1m sex       \u001b[0m\u001b[1m phenotype \u001b[0m\n",
       "\u001b[1m     \u001b[0mâ”‚\u001b[90m Abstractâ€¦ \u001b[0m\u001b[90m Abstractâ€¦ \u001b[0m\u001b[90m Abstractâ€¦ \u001b[0m\u001b[90m Abstractâ€¦ \u001b[0m\u001b[90m Abstractâ€¦ \u001b[0m\u001b[90m Abstractâ€¦ \u001b[0m\n",
       "â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
       "  20 â”‚ 0          HG00119    0          0          0          -9"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 20\n",
    "d.person_info[i, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e40f976b",
   "metadata": {},
   "source": [
    "Information of j-th variant is accessible by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "88ff4de4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<p>DataFrameRow (6 columns)</p><div class=\"data-frame\"><table class=\"data-frame\"><thead><tr><th></th><th>chromosome</th><th>snpid</th><th>genetic_distance</th><th>position</th><th>allele1</th><th>allele2</th></tr><tr><th></th><th title=\"String\">String</th><th title=\"String\">String</th><th title=\"Float64\">Float64</th><th title=\"Int64\">Int64</th><th title=\"String\">String</th><th title=\"String\">String</th></tr></thead><tbody><tr><th>7</th><td>22</td><td>rs114690707</td><td>0.0</td><td>20000771</td><td>C</td><td>A</td></tr></tbody></table></div>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccc}\n",
       "\t& chromosome & snpid & genetic\\_distance & position & allele1 & allele2\\\\\n",
       "\t\\hline\n",
       "\t& String & String & Float64 & Int64 & String & String\\\\\n",
       "\t\\hline\n",
       "\t7 & 22 & rs114690707 & 0.0 & 20000771 & C & A \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1mDataFrameRow\u001b[0m\n",
       "\u001b[1m Row \u001b[0mâ”‚\u001b[1m chromosome \u001b[0m\u001b[1m snpid       \u001b[0m\u001b[1m genetic_distance \u001b[0m\u001b[1m position \u001b[0m\u001b[1m allele1 \u001b[0m\u001b[1m allele2 \u001b[0m\n",
       "\u001b[1m     \u001b[0mâ”‚\u001b[90m String     \u001b[0m\u001b[90m String      \u001b[0m\u001b[90m Float64          \u001b[0m\u001b[90m Int64    \u001b[0m\u001b[90m String  \u001b[0m\u001b[90m String  \u001b[0m\n",
       "â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
       "   7 â”‚ 22          rs114690707               0.0  20000771  C        A"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "j = 7\n",
    "d.snp_info[j, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82e7aae2",
   "metadata": {},
   "source": [
    "Genotype of sample $i$, variant $j$ is accessible by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a4a688fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0x03"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.snparray[i, j]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bfc3462",
   "metadata": {},
   "source": [
    "Note that this is the encoding defined by the table above. If converted numerically in `0`, `1`, `2`-encoding, it corresponds to `2`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1740c4b5",
   "metadata": {},
   "source": [
    "Genotypes of each variant in numeric form can be accessed by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c617b424",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0\n",
      "2.0\n",
      "2.0\n",
      "2.0\n",
      "1.706806282722513\n",
      "2.0\n",
      "2.0\n",
      "1.9842931937172774\n",
      "2.0\n",
      "2.0\n",
      "1.9947643979057592\n",
      "2.0\n",
      "0.06806282722513089\n",
      "2.0\n",
      "2.0\n",
      "0.19895287958115182\n",
      "1.9947643979057592\n",
      "2.0\n",
      "2.0\n",
      "1.9842931937172774\n",
      "2.0\n",
      "1.963350785340314\n",
      "2.0\n",
      "2.0\n",
      "2.0\n",
      "2.0\n",
      "2.0\n",
      "1.9528795811518325\n",
      "2.0\n",
      "2.0\n"
     ]
    }
   ],
   "source": [
    "g = Vector{Float64}(undef, d.people)\n",
    "cnt = 0\n",
    "for j in 1:30\n",
    "    copyto!(g, @view(d.snparray[:, j]); impute=true, model=ADDITIVE_MODEL, center=false, scale=false)\n",
    "    println(mean(g))\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126e9e34",
   "metadata": {},
   "source": [
    "Note that the count of the second allele is encoded, and it is often the reference allele. If it is desired to run the analyses based on the alternate allele count, the values have to be reversed (subtracted from `2.0`). This led to some projects putting the reference allele first, most notably UK Biobank. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af35f9b7",
   "metadata": {},
   "source": [
    "### Oxford BGEN format (`.bgen` + optional `.sample`)\n",
    "The BGEN format is native to Oxford statistical genetics tools, such as [IMPUTE2](https://mathgen.stats.ox.ac.uk/impute/impute_v2.html) and [SNPTEST](https://www.well.ox.ac.uk/~gav/snptest/). This format employs variant-by-variant compression scheme, well-tailored for GWAS applications. The UK Biobank imputed data is distributed in this format. \n",
    "\n",
    "- __Pros__: \n",
    "    - Very small file size. Further compression from the fixed-width genotype representation\n",
    "    - Admits variable precision.\n",
    "    - Variant-by-variant compression suitable for GWAS, with dosage and phase information supported\n",
    "    - Allows dosage and phase information\n",
    "- __Cons__: \n",
    "    - Needs an external index for random access\n",
    "    - Compression not tailored for genetic context\n",
    "    - REF/ALT designation unsupported\n",
    "    \n",
    "#### Basic usage (`BGEN.jl`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "87f3d7d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: using BGEN.samples in module Main conflicts with an existing identifier.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Bgen(IOStream(<file test_bgen.bgen>), 0x0000000000020c71, BGEN.Header(0x000006d3, 0x00000014, 0x0000054c, 0x000000bf, 0x02, 0x02, false), [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\"  â€¦  \"182\", \"183\", \"184\", \"185\", \"186\", \"187\", \"188\", \"189\", \"190\", \"191\"], nothing)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "using BGEN\n",
    "b = Bgen(\"test_bgen.bgen\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39a99bca",
   "metadata": {},
   "source": [
    "Sample names are accessible by: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fc0b5dc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "191-element Vector{String}:\n",
       " \"1\"\n",
       " \"2\"\n",
       " \"3\"\n",
       " \"4\"\n",
       " \"5\"\n",
       " \"6\"\n",
       " \"7\"\n",
       " \"8\"\n",
       " \"9\"\n",
       " \"10\"\n",
       " \"11\"\n",
       " \"12\"\n",
       " \"13\"\n",
       " â‹®\n",
       " \"180\"\n",
       " \"181\"\n",
       " \"182\"\n",
       " \"183\"\n",
       " \"184\"\n",
       " \"185\"\n",
       " \"186\"\n",
       " \"187\"\n",
       " \"188\"\n",
       " \"189\"\n",
       " \"190\"\n",
       " \"191\""
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BGEN.samples(b)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75ca743f",
   "metadata": {},
   "source": [
    "Since how to iterate across the variants in BGEN files depend on if the index file is provided, we support an iterator interface for variants. If you are familiar with Python, it is an interface similar to generator defined using `yield` statement in Python. This iterator is accessible by the function `iterator()`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8dc9d0b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rsid\tchrom\tpos\tn_alleles\tlist of alleles\n",
      "rs138720731\t22\t20000086\t2\t[\"C\", \"T\"]\n",
      "rs73387790\t22\t20000146\t2\t[\"A\", \"G\"]\n",
      "rs183293480\t22\t20000199\t2\t[\"C\", \"A\"]\n",
      "rs185807825\t22\t20000291\t2\t[\"T\", \"G\"]\n",
      "rs55902548\t22\t20000428\t2\t[\"T\", \"G\"]\n",
      "rs142720028\t22\t20000683\t2\t[\"G\", \"A\"]\n",
      "rs114690707\t22\t20000771\t2\t[\"C\", \"A\"]\n",
      "rs189842693\t22\t20000793\t2\t[\"C\", \"T\"]\n",
      "rs147349046\t22\t20000810\t2\t[\"T\", \"C\"]\n",
      "rs183154520\t22\t20000814\t2\t[\"C\", \"T\"]\n",
      "rs187930998\t22\t20000864\t2\t[\"A\", \"G\"]\n",
      "rs148068532\t22\t20000882\t2\t[\"G\", \"C\"]\n",
      "rs1978233\t22\t20000950\t2\t[\"G\", \"T\"]\n",
      "rs141800233\t22\t20000975\t2\t[\"A\", \"G\"]\n",
      "rs192051979\t22\t20001001\t2\t[\"C\", \"T\"]\n",
      "rs2079702\t22\t20001006\t2\t[\"A\", \"G\"]\n",
      "rs183256914\t22\t20001016\t2\t[\"T\", \"C\"]\n",
      "rs150580380\t22\t20001157\t2\t[\"A\", \"G\"]\n",
      "rs139570132\t22\t20001159\t2\t[\"T\", \"C\"]\n",
      "rs143369598\t22\t20001219\t2\t[\"C\", \"G\"]\n",
      "rs5993894\t22\t20001333\t2\t[\"T\", \"C\"]\n",
      "rs146344141\t22\t20001434\t2\t[\"T\", \"C\"]\n",
      "rs188666449\t22\t20001455\t2\t[\"A\", \"G\"]\n",
      "rs139601437\t22\t20001521\t2\t[\"A\", \"C\"]\n",
      "rs71788814\t22\t20001587\t2\t[\"C\", \"CAG\"]\n",
      "rs144217522\t22\t20001600\t2\t[\"A\", \"T\"]\n",
      "rs192606530\t22\t20001655\t2\t[\"A\", \"G\"]\n",
      "rs111598545\t22\t20001822\t2\t[\"G\", \"A\"]\n",
      "rs184950746\t22\t20002011\t2\t[\"T\", \"C\"]\n",
      "rs142461772\t22\t20002207\t2\t[\"A\", \"G\"]\n"
     ]
    }
   ],
   "source": [
    "println(\"rsid\\tchrom\\tpos\\tn_alleles\\tlist of alleles\")\n",
    "cnt = 0\n",
    "for v in iterator(b)\n",
    "    println(\"$(rsid(v::Variant))\\t$(chrom(v))\\t$(pos(v))\\t$(n_alleles(v))\\t$(alleles(v))\")\n",
    "    cnt += 1 \n",
    "    if cnt == 30\n",
    "        break\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13aecf82",
   "metadata": {},
   "source": [
    "Dosage of each variant is accssible by: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c353baac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.29319373\n",
      "0.0\n",
      "0.0\n",
      "0.015706806\n",
      "0.0\n",
      "0.0\n",
      "0.005235602\n",
      "0.0\n",
      "1.9319372\n",
      "0.0\n",
      "0.0\n",
      "1.8010471\n",
      "0.005235602\n",
      "0.0\n",
      "0.0\n",
      "0.015706806\n",
      "0.0\n",
      "0.036649216\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.04712042\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "cnt = 0\n",
    "for v in BGEN.iterator(b)\n",
    "    g = first_allele_dosage!(b, v) # first allele is the ALT allele in this file.\n",
    "    println(mean(g))\n",
    "    cnt += 1\n",
    "    if cnt == 30\n",
    "        break\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "983134f0",
   "metadata": {},
   "source": [
    "### PLINK 2 PGEN format (`.pgen` + `.pvar` and `.psam`)\n",
    "\n",
    "PGEN is a backward-compatible extension of the BED format for [PLINK 2](https://www.cog-genomics.org/plink/2.0/) under development. It tries to overcome the limitation of the BED format, and can incorporate phase and dosage information. Cutting-edge GWAS tools now support this format. \n",
    "\n",
    "- __Pros__: \n",
    "    - Very small file size, similar to BGEN.\n",
    "    - Variant-by-variant compression suitable for GWAS. \n",
    "    - Allows dosage and phase information\n",
    "    - Difflist or patch list: genetics-inspired compression incorporating linkage disequillibrium\n",
    "        - 4x Faster than reading in `BGEN` with similar precision\n",
    "- __Cons__: \n",
    "    - Still under development\n",
    "    - Difficult to write a parser by admitting too many forms, including the whole BED format.\n",
    "    - Requires external variant (`.pvar`) and sample (`.psam`) files. \n",
    "    \n",
    "Note: [Linkage disequilibrium](https://www.sciencedirect.com/topics/neuroscience/linkage-disequilibrium)\n",
    "\n",
    "> Linkage disequilibrium (LD) is the correlation between nearby variants such that the alleles at neighboring polymorphisms (observed on the same chromosome) are associated within a population more often than if they were unlinked.\n",
    "\n",
    "#### Basic usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9ec6c979",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: using PGENFiles.iterator in module Main conflicts with an existing identifier.\n"
     ]
    }
   ],
   "source": [
    "using PGENFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2dcaca10",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = Pgen(\"test_pgen.pgen\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a3fd819d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1356"
     ]
    }
   ],
   "source": [
    "print(PGENFiles.n_variants(p))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bad153f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "191"
     ]
    }
   ],
   "source": [
    "print(PGENFiles.n_samples(p))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fc7c9ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_iter = PGENFiles.iterator(p);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1312a521",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(UInt8[0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00  â€¦  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], UInt8[0x00], 0x0000000000000001)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v = first(v_iter)\n",
    "g, data, offset = get_genotypes(p, v)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2fbe42b",
   "metadata": {},
   "source": [
    "`g` is the genotype vector, `data` is the variant record, and `offset` denotes the end of dosage record on the current variant record track."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "49e3d9d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "191-element Vector{UInt8}:\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       "    â‹®\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00\n",
       " 0x00"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1722d11",
   "metadata": {},
   "source": [
    "The encoding is as follows:\n",
    "\n",
    "| genotype code |\tgenotype category |\n",
    "|:---:|:---:|\n",
    "| `0x00` | \thomozygous REF |\n",
    "| `0x01` |\theterozygous REF-ALT |\n",
    "| `0x02` |\thomozygous ALT |\n",
    "| `0x03` |\tmissing |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f23f84c9",
   "metadata": {},
   "source": [
    "To avoid array allocations for iterative genotype extraction, one may preallocate `g` and reuse it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8144feb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.2931937172774869\n",
      "0.0\n",
      "0.0\n",
      "0.015706806282722512\n",
      "0.0\n",
      "0.0\n",
      "0.005235602094240838\n",
      "0.0\n",
      "1.931937172774869\n",
      "0.0\n",
      "0.0\n",
      "1.801047120418848\n",
      "0.005235602094240838\n",
      "0.0\n",
      "0.0\n",
      "0.015706806282722512\n",
      "0.0\n",
      "0.03664921465968586\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.04712041884816754\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "g = Vector{UInt8}(undef, PGENFiles.n_samples(p))\n",
    "cnt = 0\n",
    "for v in PGENFiles.iterator(p)\n",
    "    get_genotypes!(g, p, v)\n",
    "    println(mean(g))\n",
    "    # do someting with genotypes in `g`...\n",
    "    cnt += 1\n",
    "    if cnt == 30\n",
    "        break\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bec1602",
   "metadata": {},
   "source": [
    "Similarly, ALT allele dosages are available through the functions `alt_allele_dosage()` and `alt_allele_dosage!()`. As genotype information is required to retrieve dosages, space for genotypes are also required for `alt_allele_dosage!()`. These functions return dosages, parsed genotypes, and endpoint of the dosage information in the current variant record."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7139e283",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.29319373\n",
      "0.0\n",
      "0.0\n",
      "0.015706806\n",
      "0.0\n",
      "0.0\n",
      "0.005235602\n",
      "0.0\n",
      "1.9319372\n",
      "0.0\n",
      "0.0\n",
      "1.8010471\n",
      "0.005235602\n",
      "0.0\n",
      "0.0\n",
      "0.015706806\n",
      "0.0\n",
      "0.036649216\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.04712042\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "d = Vector{Float32}(undef, PGENFiles.n_samples(p))\n",
    "g = Vector{UInt8}(undef, PGENFiles.n_samples(p))\n",
    "g_ld = similar(g)\n",
    "cnt = 0\n",
    "for v in v_iter\n",
    "    cnt += 1\n",
    "    alt_allele_dosage!(d, g, p, v)\n",
    "    println(mean(d))\n",
    "    # do someting with dosage values in `d`...\n",
    "    if cnt == 30\n",
    "        break\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "387a26d9",
   "metadata": {},
   "source": [
    "Information of each sample and variant is available by reading in `.psam` and `.pvar` file as a `DataFrame` (not yet supported by `PGENFiles.jl`). `.pvar` format admits regular `.vcf` format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "37d6aa7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "using DataFrames, CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5d479456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"data-frame\"><p>5 rows Ã— 2 columns</p><table class=\"data-frame\"><thead><tr><th></th><th>#IID</th><th>SEX</th></tr><tr><th></th><th title=\"String7\">String7</th><th title=\"String3\">String3</th></tr></thead><tbody><tr><th>1</th><td>HG00096</td><td>NA</td></tr><tr><th>2</th><td>HG00097</td><td>NA</td></tr><tr><th>3</th><td>HG00099</td><td>NA</td></tr><tr><th>4</th><td>HG00100</td><td>NA</td></tr><tr><th>5</th><td>HG00101</td><td>NA</td></tr></tbody></table></div>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cc}\n",
       "\t& \\#IID & SEX\\\\\n",
       "\t\\hline\n",
       "\t& String7 & String3\\\\\n",
       "\t\\hline\n",
       "\t1 & HG00096 & NA \\\\\n",
       "\t2 & HG00097 & NA \\\\\n",
       "\t3 & HG00099 & NA \\\\\n",
       "\t4 & HG00100 & NA \\\\\n",
       "\t5 & HG00101 & NA \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1m5Ã—2 DataFrame\u001b[0m\n",
       "\u001b[1m Row \u001b[0mâ”‚\u001b[1m #IID    \u001b[0m\u001b[1m SEX     \u001b[0m\n",
       "\u001b[1m     \u001b[0mâ”‚\u001b[90m String7 \u001b[0m\u001b[90m String3 \u001b[0m\n",
       "â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
       "   1 â”‚ HG00096  NA\n",
       "   2 â”‚ HG00097  NA\n",
       "   3 â”‚ HG00099  NA\n",
       "   4 â”‚ HG00100  NA\n",
       "   5 â”‚ HG00101  NA"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info = CSV.read(\"test_pgen.psam\", DataFrame)\n",
    "first(sample_info, 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "889b950a",
   "metadata": {},
   "source": [
    "For `.pvar` file, all the lines starting with `#` is header, ending with the column names. For `test_pgen.pvar`, it is the 26th line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8f71271f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"data-frame\"><p>5 rows Ã— 8 columns</p><table class=\"data-frame\"><thead><tr><th></th><th>#CHROM</th><th>POS</th><th>ID</th><th>REF</th><th>ALT</th><th>QUAL</th><th>FILTER</th><th>INFO</th></tr><tr><th></th><th title=\"Int64\">Int64</th><th title=\"Int64\">Int64</th><th title=\"String15\">String15</th><th title=\"String15\">String15</th><th title=\"String31\">String31</th><th title=\"String7\">String7</th><th title=\"String7\">String7</th><th title=\"String\">String</th></tr></thead><tbody><tr><th>1</th><td>22</td><td>20000086</td><td>rs138720731</td><td>T</td><td>C</td><td>100</td><td>PASS</td><td>AC=7;RSQ=0.8454;AVGPOST=0.9983;AA=T;AN=2184;LDAF=0.0040;THETA=0.0001;VT=SNP;SNPSOURCE=LOWCOV;ERATE=0.0003;AF=0.0032;AFR_AF=0.01</td></tr><tr><th>2</th><td>22</td><td>20000146</td><td>rs73387790</td><td>G</td><td>A</td><td>100</td><td>PASS</td><td>LDAF=0.0169;RSQ=0.9482;THETA=0.0004;AA=G;AN=2184;AVGPOST=0.9972;VT=SNP;SNPSOURCE=LOWCOV;AC=36;ERATE=0.0003;AF=0.02;AFR_AF=0.07;EUR_AF=0.0013</td></tr><tr><th>3</th><td>22</td><td>20000199</td><td>rs183293480</td><td>A</td><td>C</td><td>100</td><td>PASS</td><td>LDAF=0.0009;THETA=0.0004;AN=2184;AVGPOST=0.9990;VT=SNP;AA=A;RSQ=0.6274;SNPSOURCE=LOWCOV;AC=1;ERATE=0.0003;AF=0.0005;EUR_AF=0.0013</td></tr><tr><th>4</th><td>22</td><td>20000291</td><td>rs185807825</td><td>G</td><td>T</td><td>100</td><td>PASS</td><td>ERATE=0.0005;AVGPOST=0.9983;AA=G;AN=2184;LDAF=0.0015;VT=SNP;SNPSOURCE=LOWCOV;RSQ=0.5564;AC=2;THETA=0.0003;AF=0.0009;ASN_AF=0.0035</td></tr><tr><th>5</th><td>22</td><td>20000428</td><td>rs55902548</td><td>G</td><td>T</td><td>100</td><td>PASS</td><td>AC=323;AVGPOST=0.9983;AA=G;AN=2184;VT=SNP;RSQ=0.9949;LDAF=0.1473;SNPSOURCE=LOWCOV;ERATE=0.0003;THETA=0.0003;AF=0.15;ASN_AF=0.0017;AMR_AF=0.15;AFR_AF=0.31;EUR_AF=0.15</td></tr></tbody></table></div>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccccc}\n",
       "\t& \\#CHROM & POS & ID & REF & ALT & QUAL & FILTER & INFO\\\\\n",
       "\t\\hline\n",
       "\t& Int64 & Int64 & String15 & String15 & String31 & String7 & String7 & String\\\\\n",
       "\t\\hline\n",
       "\t1 & 22 & 20000086 & rs138720731 & T & C & 100 & PASS & AC=7;RSQ=0.8454;AVGPOST=0.9983;AA=T;AN=2184;LDAF=0.0040;THETA=0.0001;VT=SNP;SNPSOURCE=LOWCOV;ERATE=0.0003;AF=0.0032;AFR\\_AF=0.01 \\\\\n",
       "\t2 & 22 & 20000146 & rs73387790 & G & A & 100 & PASS & LDAF=0.0169;RSQ=0.9482;THETA=0.0004;AA=G;AN=2184;AVGPOST=0.9972;VT=SNP;SNPSOURCE=LOWCOV;AC=36;ERATE=0.0003;AF=0.02;AFR\\_AF=0.07;EUR\\_AF=0.0013 \\\\\n",
       "\t3 & 22 & 20000199 & rs183293480 & A & C & 100 & PASS & LDAF=0.0009;THETA=0.0004;AN=2184;AVGPOST=0.9990;VT=SNP;AA=A;RSQ=0.6274;SNPSOURCE=LOWCOV;AC=1;ERATE=0.0003;AF=0.0005;EUR\\_AF=0.0013 \\\\\n",
       "\t4 & 22 & 20000291 & rs185807825 & G & T & 100 & PASS & ERATE=0.0005;AVGPOST=0.9983;AA=G;AN=2184;LDAF=0.0015;VT=SNP;SNPSOURCE=LOWCOV;RSQ=0.5564;AC=2;THETA=0.0003;AF=0.0009;ASN\\_AF=0.0035 \\\\\n",
       "\t5 & 22 & 20000428 & rs55902548 & G & T & 100 & PASS & AC=323;AVGPOST=0.9983;AA=G;AN=2184;VT=SNP;RSQ=0.9949;LDAF=0.1473;SNPSOURCE=LOWCOV;ERATE=0.0003;THETA=0.0003;AF=0.15;ASN\\_AF=0.0017;AMR\\_AF=0.15;AFR\\_AF=0.31;EUR\\_AF=0.15 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1m5Ã—8 DataFrame\u001b[0m\n",
       "\u001b[1m Row \u001b[0mâ”‚\u001b[1m #CHROM \u001b[0m\u001b[1m POS      \u001b[0m\u001b[1m ID          \u001b[0m\u001b[1m REF      \u001b[0m\u001b[1m ALT      \u001b[0m\u001b[1m QUAL    \u001b[0m\u001b[1m FILTER  \u001b[0m\u001b[1m IN\u001b[0m â‹¯\n",
       "\u001b[1m     \u001b[0mâ”‚\u001b[90m Int64  \u001b[0m\u001b[90m Int64    \u001b[0m\u001b[90m String15    \u001b[0m\u001b[90m String15 \u001b[0m\u001b[90m String31 \u001b[0m\u001b[90m String7 \u001b[0m\u001b[90m String7 \u001b[0m\u001b[90m St\u001b[0m â‹¯\n",
       "â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
       "   1 â”‚     22  20000086  rs138720731  T         C         100      PASS     AC â‹¯\n",
       "   2 â”‚     22  20000146  rs73387790   G         A         100      PASS     LD\n",
       "   3 â”‚     22  20000199  rs183293480  A         C         100      PASS     LD\n",
       "   4 â”‚     22  20000291  rs185807825  G         T         100      PASS     ER\n",
       "   5 â”‚     22  20000428  rs55902548   G         T         100      PASS     AC â‹¯\n",
       "\u001b[36m                                                                1 column omitted\u001b[0m"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_info = CSV.read(\"test_pgen.pvar\", DataFrame; delim=\"\\t\", header=26)\n",
    "first(variant_info, 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aee7d153",
   "metadata": {},
   "source": [
    "## Exercise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bd20aa",
   "metadata": {},
   "source": [
    "[Minor allele frequency](https://en.wikipedia.org/wiki/Minor_allele_frequency) (MAF) is widely used for determining if a variant is rare or frequent, and in GWAS, it is used for measuring information content of a variant. One example is the ratio of actual numerical variance and expected variance from the binomial model ($2 \\hat{p}(1-\\hat{p})$, where $\\hat{p}$ is the MAF). How would you compute minor allele frequencies using the packages we used? How would you determine the minor/major allele? How would you compute the variance ratio measure?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6b7249b",
   "metadata": {},
   "source": [
    "## File format transformation\n",
    "\n",
    "The file formats can be transformed between each other using plink2 on the command line. For example, the files used in this workshop is transformed from a VCF file using the following commands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8521d1a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a2.3 64-bit (24 Jan 2020)           www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to test_bgen.log.\n",
      "Options in effect:\n",
      "  --export bgen-1.3\n",
      "  --out test_bgen\n",
      "  --vcf test_vcf.vcf.gz\n",
      "\n",
      "Start time: Wed Jul 13 14:23:24 2022\n",
      "16384 MiB RAM detected; reserving 8192 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "\r",
      "--vcf: 1k variants scanned.\r",
      "--vcf: 1356 variants scanned.\n",
      "\r",
      "--vcf: 0k variants converted.    \b\b\b\b\r",
      "--vcf: test_bgen-temporary.pgen + test_bgen-temporary.pvar +\n",
      "test_bgen-temporary.psam written.\n",
      "191 samples (0 females, 0 males, 191 ambiguous; 191 founders) loaded from\n",
      "test_bgen-temporary.psam.\n",
      "1356 variants loaded from test_bgen-temporary.pvar.\n",
      "Note: No phenotype data present.\n",
      "Writing test_bgen.bgen ... 0%\b\bdone.\n",
      "Writing test_bgen.sample ... done.\n",
      "End time: Wed Jul 13 14:23:24 2022\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Process(`\u001b[4mplink2\u001b[24m \u001b[4m--vcf\u001b[24m \u001b[4mtest_vcf.vcf.gz\u001b[24m \u001b[4m--export\u001b[24m \u001b[4mbgen-1.3\u001b[24m \u001b[4m--out\u001b[24m \u001b[4mtest_bgen\u001b[24m`, ProcessExited(0))"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run(`plink2 --vcf test_vcf.vcf.gz --export bgen-1.3 --out test_bgen`) # ALT allele comes first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "06318266",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a2.3 64-bit (24 Jan 2020)           www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to test_bed.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --out test_bed\n",
      "  --vcf test_vcf.vcf.gz\n",
      "\n",
      "Start time: Wed Jul 13 14:23:24 2022\n",
      "16384 MiB RAM detected; reserving 8192 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "\r",
      "--vcf: 1k variants scanned.\r",
      "--vcf: 1356 variants scanned.\n",
      "\r",
      "--vcf: 0k variants converted.    \b\b\b\b\r",
      "--vcf: test_bed-temporary.pgen + test_bed-temporary.pvar +\n",
      "test_bed-temporary.psam written.\n",
      "191 samples (0 females, 0 males, 191 ambiguous; 191 founders) loaded from\n",
      "test_bed-temporary.psam.\n",
      "1356 variants loaded from test_bed-temporary.pvar.\n",
      "Note: No phenotype data present.\n",
      "Writing test_bed.fam ... done.\n",
      "Writing test_bed.bim ... done.\n",
      "Writing test_bed.bed ... 0%\b\bdone.\n",
      "End time: Wed Jul 13 14:23:25 2022\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Process(`\u001b[4mplink2\u001b[24m \u001b[4m--vcf\u001b[24m \u001b[4mtest_vcf.vcf.gz\u001b[24m \u001b[4m--make-bed\u001b[24m \u001b[4m--out\u001b[24m \u001b[4mtest_bed\u001b[24m`, ProcessExited(0))"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run(`plink2 --vcf test_vcf.vcf.gz --make-bed --out test_bed`) # ALT allele comes first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "146a9b54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a2.3 64-bit (24 Jan 2020)           www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to test_pgen.log.\n",
      "Options in effect:\n",
      "  --make-pgen\n",
      "  --out test_pgen\n",
      "  --vcf test_vcf.vcf.gz\n",
      "\n",
      "Start time: Wed Jul 13 14:23:25 2022\n",
      "16384 MiB RAM detected; reserving 8192 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "\r",
      "--vcf: 1k variants scanned.\r",
      "--vcf: 1356 variants scanned.\n",
      "\r",
      "--vcf: 0k variants converted.    \b\b\b\b\r",
      "--vcf: test_pgen-temporary.pgen + test_pgen-temporary.pvar +\n",
      "test_pgen-temporary.psam written.\n",
      "191 samples (0 females, 0 males, 191 ambiguous; 191 founders) loaded from\n",
      "test_pgen-temporary.psam.\n",
      "1356 variants loaded from test_pgen-temporary.pvar.\n",
      "Note: No phenotype data present.\n",
      "Writing test_pgen.psam ... done.\n",
      "Writing test_pgen.pvar ... 0%\b\b0%\b\b1%\b\b1%\b\b2%\b\b2%\b\b3%\b\b3%\b\b4%\b\b4%\b\b5%\b\b5%\b\b6%\b\b6%\b\b7%\b\b7%\b\b8%\b\b8%\b\b9%\b\b9%\b\b10%\b\b\b10%\b\b\b11%\b\b\b11%\b\b\b12%\b\b\b12%\b\b\b13%\b\b\b13%\b\b\b14%\b\b\b14%\b\b\b15%\b\b\b15%\b\b\b16%\b\b\b16%\b\b\b17%\b\b\b17%\b\b\b18%\b\b\b18%\b\b\b19%\b\b\b19%\b\b\b20%\b\b\b20%\b\b\b21%\b\b\b21%\b\b\b22%\b\b\b22%\b\b\b23%\b\b\b23%\b\b\b24%\b\b\b25%\b\b\b25%\b\b\b26%\b\b\b26%\b\b\b27%\b\b\b27%\b\b\b28%\b\b\b28%\b\b\b29%\b\b\b29%\b\b\b30%\b\b\b30%\b\b\b31%\b\b\b31%\b\b\b32%\b\b\b32%\b\b\b33%\b\b\b33%\b\b\b34%\b\b\b34%\b\b\b35%\b\b\b35%\b\b\b36%\b\b\b36%\b\b\b37%\b\b\b37%\b\b\b38%\b\b\b38%\b\b\b39%\b\b\b39%\b\b\b40%\b\b\b40%\b\b\b41%\b\b\b41%\b\b\b42%\b\b\b42%\b\b\b43%\b\b\b43%\b\b\b44%\b\b\b44%\b\b\b45%\b\b\b45%\b\b\b46%\b\b\b46%\b\b\b47%\b\b\b47%\b\b\b48%\b\b\b48%\b\b\b49%\b\b\b50%\b\b\b50%\b\b\b51%\b\b\b51%\b\b\b52%\b\b\b52%\b\b\b53%\b\b\b53%\b\b\b54%\b\b\b54%\b\b\b55%\b\b\b55%\b\b\b56%\b\b\b56%\b\b\b57%\b\b\b57%\b\b\b58%\b\b\b58%\b\b\b59%\b\b\b59%\b\b\b60%\b\b\b60%\b\b\b61%\b\b\b61%\b\b\b62%\b\b\b62%\b\b\b63%\b\b\b63%\b\b\b64%\b\b\b64%\b\b\b65%\b\b\b65%\b\b\b66%\b\b\b66%\b\b\b67%\b\b\b67%\b\b\b68%\b\b\b68%\b\b\b69%\b\b\b69%\b\b\b70%\b\b\b70%\b\b\b71%\b\b\b71%\b\b\b72%\b\b\b72%\b\b\b73%\b\b\b73%\b\b\b74%\b\b\b75%\b\b\b75%\b\b\b76%\b\b\b76%\b\b\b77%\b\b\b77%\b\b\b78%\b\b\b78%\b\b\b79%\b\b\b79%\b\b\b80%\b\b\b80%\b\b\b81%\b\b\b81%\b\b\b82%\b\b\b82%\b\b\b83%\b\b\b83%\b\b\b84%\b\b\b84%\b\b\b85%\b\b\b85%\b\b\b86%\b\b\b86%\b\b\b87%\b\b\b87%\b\b\b88%\b\b\b88%\b\b\b89%\b\b\b89%\b\b\b90%\b\b\b90%\b\b\b91%\b\b\b91%\b\b\b92%\b\b\b92%\b\b\b93%\b\b\b93%\b\b\b94%\b\b\b94%\b\b\b95%\b\b\b95%\b\b\b96%\b\b\b96%\b\b\b97%\b\b\b97%\b\b\b98%\b\b\b98%\b\b\b99%\b\b\bdone.\n",
      "Writing test_pgen.pgen ... 0%\b\bdone.\n",
      "End time: Wed Jul 13 14:23:25 2022\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Process(`\u001b[4mplink2\u001b[24m \u001b[4m--vcf\u001b[24m \u001b[4mtest_vcf.vcf.gz\u001b[24m \u001b[4m--make-pgen\u001b[24m \u001b[4m--out\u001b[24m \u001b[4mtest_pgen\u001b[24m`, ProcessExited(0))"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run(`plink2 --vcf test_vcf.vcf.gz --make-pgen --out test_pgen`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65007448",
   "metadata": {},
   "source": [
    "## Concluding Remarks: Applications in OpenMendel\n",
    "\n",
    "While we focus on variant-by-variant access of the genetic variant data, many of the packages we have seen has various utility functionalities such as filtering out the variants with low MAF or low genotype success rate, filter by chromosome, and merge. In case of the BED format (`SnpArrays.jl`), we support high-performance linear algebra on the genotype matrix which supports multithreading and GPU computation. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cff4a1f5",
   "metadata": {},
   "source": [
    "### GWAS application\n",
    "\n",
    "- `OrdinalGWAS.jl` : GWAS for ordered categorical trait, e.g. disease status (undiagnosed, pre-disease, mild, moderate, severe)\n",
    "- `TrajGWAS.jl` : GWAS for continuous longitudinal phenotypes using a modified linear mixed effects model. Tests both mean effect and within-subject variability effect.\n",
    "\n",
    "\n",
    "### High-performance applications using fixed-width BED format\n",
    "\n",
    "- `MendelIHT.jl` : Rather than GWAS with variant-by-variant testing, uses penalized regression model. \n",
    "- `OpenADMIXTURE.jl` : Estimates ancestry of samples. Julia reimplementation of highly celebrated ADMIXTURE software in C++, 8x faster in multi-threaded setting and 35x faster using GPU.\n",
    "\n",
    "\n",
    "### Others\n",
    "- `VarianceComponentModels.jl` : Fitting and testing variance component models\n",
    "- `MendelImpute.jl` : Fast phase inference and genotype imputation\n",
    "- `TraitSimulation.jl` : Quickly simulate phenotypes under a variety of genetic architectures.\n",
    "- `QuasiCopula.jl` : Analysis of correlated data with specified marginals with a flexible quasi-copula distribution"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.7.1",
   "language": "julia",
   "name": "julia-1.7"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
